[SR] Snapshot and restore plugin boilerplate

jen-huang opened the pull request:
The beginning of snapshot and restore GUI!

This PR sets up the new plugin and includes some boilerplate client-side and server-side code. Since this plugin was created with [new Kibana platform principles](https://github.com/elastic/kibana/pull/31147) in mind, I am primarily looking for feedback with regards to that. You'll notice two sets of `shim.ts` and `plugin.ts` files, one used by client and the other for server.

@epixa, I ran into some difficulties with accessing `core` and `plugins` from child React components. I ended up creating a React context, [`AppContext`](https://github.com/elastic/kibana/pull/32276/files#diff-e9abe9d8bf3c2fe2da9d6ef3341887d9), which is initialized when [parent component is mounted](https://github.com/elastic/kibana/pull/32276/files#diff-8d4511e4baffc9cb7ef386b221f0eab0R27), and referenced by [child components using `contextType`](https://github.com/elastic/kibana/pull/32276/files#diff-6082f371bbc8b4993eb4529f8aa8ea71R19) so that `core` and `plugins` can be used in lifecycle methods. Feedback and/or suggestions for this would be super appreciated.

elasticmachine commented:
Pinging @elastic/es-ui

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/3106/)

cjcenizal commented on x-pack/plugins/snapshot_restore/server/routes/api/repositories.ts:
```diff
@@ -0,0 +1,36 @@
+export const getAll = {
+  path: 'repositories',
+  method: 'get',
+  handler: async (req: any, callWithRequest: any, responseToolkit: any) => {
```
From looking at how some other plugins do it, I think we can use a few types here:

```js
import { Legacy } from 'kibana';
import { Request, ResponseToolkit } from 'hapi';
import { CallClusterWithRequest } from 'src/legacy/core_plugins/elasticsearch';

handler: async (req: Legacy.Request, callWithRequest: CallClusterWithRequest, responseToolkit: ResponseToolkit) => {

// or this one... not sure why we have both Legacy.Request and Request
handler: async (req:Request, callWithRequest: CallClusterWithRequest, responseToolkit: ResponseToolkit) => {
```

cjcenizal commented on x-pack/server/lib/create_router/index.d.ts:
```diff
@@ -0,0 +1,26 @@
+  patch: RouterRoute;
+}
+
+export declare function createRouter(server: any, pluginId: string, apiBasePath: string): Router;
```
We can type `server`:

```js
import { Server } from 'hapi';
export declare function createRouter(server: Server, pluginId: string, apiBasePath: string): Router;
```

cjcenizal commented on x-pack/server/lib/create_router/index.d.ts:
```diff
@@ -0,0 +1,26 @@
+
+import { Request, ResponseObject, ResponseToolkit } from 'hapi';
+
+type CallWithRequest = (request: any, action: string, params: any) => any;
```
I believe we can type `request` per my comment above.

epixa replied:
I think the legacy ES call stuff is all typed now: https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/elasticsearch/index.d.ts

cjcenizal commented on x-pack/server/lib/register_license_checker/index.d.ts:
```diff
@@ -0,0 +1,14 @@
+import { LicenseType } from '../../../common/constants';
+
+export declare function registerLicenseChecker(
+  server: any,
```
We can type `server`, per my comment above.

epixa commented on x-pack/plugins/snapshot_restore/plugin.ts:
```diff
@@ -0,0 +1,34 @@
+
+  public start(): void {
+    const server = this.server;
+    const { core, plugins } = createShim(server, PLUGIN.ID);
```
I recommend pulling the shimming logic into `snapshot_restore/index.ts`, so `plugin.ts` doesn't appear coupled to the legacy stuff at all. So you'd call `createShim` from the legacy `init()` and pass `core` and `plugins` as arguments to this `start()`.

Over time you'll convert more of the shim stuff over to new platform services directly, and then eventually you'll just delete the legacy plugin definition entirely, and your new plugin definition won't really change along the way.

epixa commented:
The react context approach makes a lot of sense in this case, I think. You might want to be more selective in passing down the features you want to use though, rather than just passing the whole arguments through.

It forces developers to make an explicit decision to take advantage of a new plugin or core feature somewhere deep within your code, which can help increase scrutiny around increased levels of coupling.

At least this is all TypeScript so it's easy to manage even if you do pass them through.

epixa commented on x-pack/plugins/snapshot_restore/public/plugin.ts:
```diff
@@ -0,0 +1,76 @@
+
+export class Plugin {
+  public start(): void {
+    const shim = createShim();
```
Like in my comment on the server, I think you should pull the shim logic out of the `Plugin` class and into the code that instantiates this and calls .start(). Keep the new code as pristine as possible, and let the legacy code handle the hackery.

epixa commented on x-pack/plugins/snapshot_restore/public/plugin.ts:
```diff
@@ -0,0 +1,76 @@
+    });
+
+    // Register react root
+    routing.registerAngularRoute(`${CLIENT_BASE_PATH}/:section?/:subsection?/:view?/:id`, {
```
Can you explain why this needs to be an angular route/controller rather than just being react like the rest of the plugin?

jen-huang replied:
since the Kibana chrome is using Angular routing, we need to register the initial base path with Angular so it's accessible from the management menu. we then mount the React app inside it which handles the rest of internal stuff in React.

epixa replied:
So if the management app itself was built using something like react router rather angular router, this wouldn't be necessary here, is that correct?

epixa commented:
This is really awesome! I left a few comments, but overall this nailed the spirit of the migration guide, and I think it turned out well.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4278/)

jen-huang commented on x-pack/plugins/snapshot_restore/server/routes/api/repositories.ts:
```diff
@@ -0,0 +1,36 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+export const getAll = {
```
I adjusted this to something similar to your suggestion - each handler is broken out into it's own named function, and registered [registerRepositoriesRoutes()](https://github.com/elastic/kibana/pull/32276/files#diff-ad1ca88f8fdfba2841b5689dd1fcfc0c). I also added additional typings to enforce the signature of each handler function

jen-huang commented on x-pack/plugins/snapshot_restore/common/constants.ts:
```diff
@@ -0,0 +1,11 @@
+
+export const PLUGIN = {
+  ID: 'snapshot_restore',
+  NAME: 'Snapshot and Restore',
```
This one was tricky since I'm trying to avoid calling `i18n` outside of `shim.ts`. In the end, I changed this to be a function, [getI18nName()](https://github.com/elastic/kibana/pull/32276/files#diff-8c3b6d3d1d873dabdcda1e4900c203b7R14), which is [called when the plugin is initialized](https://github.com/elastic/kibana/pull/32276/files#diff-8d167ecd50c7178cf6b990fa036e2c0aR32).

Also, this is called in the legacy world rather than `plugin.ts` to avoid the plugin knowing about `server`, which is needed by `registerLicenseChecker`.

@epixa, any thoughts on accessing `i18n`, or other core/plugin modules from constant files like this case?

jen-huang commented:
Thanks for the reviews @cjcenizal and @epixa! I've addressed your feedback.

I've fleshed out the typings to what I believe is a comprehensive enough point. I will add tests in the next pass when building out the actual features, so this PR is ready for another review by you two, or anyone else I have tagged. 😄

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4283/)

restrry commented on x-pack/plugins/snapshot_restore/public/shim.ts:
```diff
@@ -0,0 +1,104 @@
+import { unmountComponentAtNode } from 'react-dom';
+import { HashRouter } from 'react-router-dom';
+
+export interface AppCore {
```
when this PR or https://github.com/elastic/kibana/pull/32746 finished we need to update typings

jen-huang replied:
thanks for the heads up 😄 I am happy to update these typings when that PR is merged. work with continue for this feature in subsequent PRs into the main feature branch, so it will be easy to update since there won't be any conflicts with master

cchaos commented on x-pack/plugins/snapshot_restore/public/index.scss:
```diff
@@ -0,0 +1,13 @@
+
+// Snapshot and restore plugin styles
+
+// Prefix all styles with "snapshot" to avoid conflicts.
```
I know there aren't any actual styles in here (yet) but prefixes should just be 3 letters. I would change it now so its correct if people start adding styles.

jen-huang replied:
thanks for that info, I've updated this to be `ssr` prefix instead

bmcconaghy commented on x-pack/plugins/snapshot_restore/public/app/index.tsx:
```diff
@@ -0,0 +1,44 @@
+import { HashRouter } from 'react-router-dom';
+
+import { App } from './app';
+// import { srStore } from './store';
```
leftover commented out code

bmcconaghy commented on x-pack/plugins/snapshot_restore/public/app/index.tsx:
```diff
@@ -0,0 +1,44 @@
+export { BASE_PATH as CLIENT_BASE_PATH } from './constants';
+
+export const renderReact = async (
+  elem: Element | null,
```
why the or null here?  seems like it should always be an Element.

bmcconaghy commented on x-pack/plugins/snapshot_restore/public/app/index.tsx:
```diff
@@ -0,0 +1,44 @@
+
+  render(
+    <I18nContext>
+      {/*<Provider store={srStore}>*/}
```
more commented out code

bmcconaghy commented on x-pack/plugins/snapshot_restore/public/app/sections/home/home.tsx:
```diff
@@ -0,0 +1,138 @@
+export class SnapshotRestoreHome extends PureComponent<Props, State> {
+  public static contextType = AppContext;
+
+  public static getDerivedStateFromProps(nextProps: Props) {
```
wondering if maybe we want to start using function components + hooks instead of class ones.

jen-huang replied:
I'd like to address this with the next pass when actual features start getting built out, it's hard for me to formulate an opinion for function components with hooks instead of lifecycles when the components don't do much yet 😁 I'll try function components + hooks for new components, and refactor these class ones if it goes well!

bmcconaghy commented on x-pack/plugins/snapshot_restore/public/plugin.ts:
```diff
@@ -0,0 +1,77 @@
+        $scope.$$postDigest(() => {
+          elem = document.getElementById(REACT_ROOT_ID);
+          renderReact(
+            elem,
```
we should check elem for null and only render if not null I think, then elem above can remove `| null`

bmcconaghy commented on x-pack/plugins/snapshot_restore/public/plugin.ts:
```diff
@@ -0,0 +1,77 @@
+
+export class Plugin {
+  public start(core: Core, plugins: Plugins): void {
+    const { i18n, routing, http, chrome, notification } = core;
```
`routing` seems like maybe not the best name for this function, since it has unmount logic, which seems unrelated to routing, more like rendering to me.

jen-huang replied:
good point, I'll move unmounting logic out of shim, and keep `routing` for just routing

bmcconaghy commented on x-pack/plugins/snapshot_restore/public/shim.ts:
```diff
@@ -0,0 +1,104 @@
+export interface Core extends AppCore {
+  routing: {
+    registerAngularRoute(path: string, config: object): void;
+    unmountReactApp(elem: Element | undefined | null): void;
```
wondering if this should be broken out into rendering

bmcconaghy commented on x-pack/plugins/snapshot_restore/public/shim.ts:
```diff
@@ -0,0 +1,104 @@
+export function createShim(): { core: Core; plugins: Plugins } {
+  // This is an Angular service, which is why we use this provider pattern
+  // to access it within our React app.
+  let httpClient: any;
```
there's a type for $http `ng.IHttpService`

jen-huang commented:
Thanks for the feedback, all! I've addressed the latest batch of em.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4474/)

epixa commented:
One detail that I don't think is obvious is that all code originating from the plugin shim "hack" file will be bundled globally into _all_ application bundles. That's the nature of hacks - they are global across all applications.

The consequence of this is potential for rapid bloat of our common bundle size as more plugins shim more stuff. This is in contrast to most other uiExport-based functionality, which is only included in bundles when it is actually used by that application. In otherwords, your snapshot code will be loaded by users even when they're accessing things like timelion, ml, and apm.

This will also be true of plugin definitions in the new platform, so it was an intentional design decision.

The way we combat this problem is by having plugin developers use async imports to lazy load code only when it is necessary. This can technically be done in any extension point that allows your plugin to register an async function, and as a plugin author you should carefully consider where it makes the most sense. Too many async imports will result in poor performance as the browser has to load a ton of JavaScript files. Poorly placed async imports can potentially result in poor performance as the browser might load a bunch of JavaScript files one at a time.

Registering management sections is a _great_ place for using an async import, and I would expect this to be the expectation of all management sections in the future. Each section potentially has a ton of code powering it, but you only need to load that code when you actually navigate to that section. Any time we see this sort of situation, an async import is probably a good idea.

Taking advantage of this requires an understanding of the problem from both sides. The management app itself must provide an extension point that allows an async function to load the section. Then the downstream plugin needs to put all of its section loading logic behind an async import. It would be pretty easy for you to do the latter in this pull request, but I'm not familiar with how management handles registering of sections, and I'm not sure if it would support an async function in its current form.

In my mind, I'd expect this to look something like:

```js
// inside start() of plugin.js
management.getCategory('elasticsearch').registerSection({
  id: 'snapshots',
  name: 'Snapshots',
  async render(domElement) {
    const { loadSection } = await import('./section');
    return loadSection(domElement, /* any core/plugin stuff you need in the section itself */);
  }
});

// section.js
// most of the logic and imports related to the snapshots section in management
// would originate here, which ensures they aren't loaded until a user navigates here
export function loadSection(domElement, ...) {
  // would render the section into the domElement container
  return function unloadSection() {
    // invoked by management app when navigating away from section
    // would destroy/cleanup anything in the section that wouldn't automatically be
    // cleaned up when the management app killed the dom element
  }
}
```

I know that's _not_ how it looks today, but perhaps the chief goal of hiding the bulk of the section logic behind an async import is technically possible with the existing API? I don't know enough about the existing API to say for sure.

jen-huang commented:
Thanks for the insights, @epixa. I agree that it would be great to use async imports for management apps. Unfortunately, the management section registry right now is not that sophisticated with regards to rendering/mounting apps; it really only maintains a list of links with a few config options.

Not hearing any other objections, I'm going to merge this into the feature branch 🙂

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4908/)
